File: card-core/src/main/java/com/jean202/cardmizer/core/domain/CardPerformancePolicy.java
package com.jean202.cardmizer.core.domain;

import com.jean202.cardmizer.common.Money;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;

public record CardPerformancePolicy(CardId cardId, List<PerformanceTier> tiers, List<BenefitRule> benefitRules) {
    public CardPerformancePolicy(CardId cardId, List<PerformanceTier> tiers) {
        this(cardId, tiers, List.of());
    }

    public CardPerformancePolicy {
        Objects.requireNonNull(cardId, "cardId must not be null");
        tiers = List.copyOf(Objects.requireNonNull(tiers, "tiers must not be null"));
        benefitRules = List.copyOf(Objects.requireNonNull(benefitRules, "benefitRules must not be null"));
        if (tiers.isEmpty()) {
            throw new IllegalArgumentException("At least one performance tier is required");
        }
    }

    public Optional<PerformanceTier> nextTier(Money spentAmount) {
        Objects.requireNonNull(spentAmount, "spentAmount must not be null");
        return sortedTiers().stream()
                .filter(tier -> tier.targetAmount().compareTo(spentAmount) > 0)
                .findFirst();
    }

    public PerformanceTier highestTier() {
        return sortedTiers().get(sortedTiers().size() - 1);
    }

    public boolean isFullyAchieved(Money spentAmount) {
        return nextTier(spentAmount).isEmpty();
    }

    public Optional<BenefitQuote> estimateBenefit(
            Money amount,
            String merchantName,
            String merchantCategory,
            Set<String> paymentTags,
            Money previousMonthSpentAmount,
            List<SpendingRecord> currentMonthRecords,
            List<SpendingRecord> currentYearRecords
    ) {
        Objects.requireNonNull(amount, "amount must not be null");
        Objects.requireNonNull(merchantName, "merchantName must not be null");
        Objects.requireNonNull(merchantCategory, "merchantCategory must not be null");
        Objects.requireNonNull(previousMonthSpentAmount, "previousMonthSpentAmount must not be null");
        List<SpendingRecord> normalizedCurrentMonthRecords = List.copyOf(
                Objects.requireNonNull(currentMonthRecords, "currentMonthRecords must not be null")
        );
        List<SpendingRecord> normalizedCurrentYearRecords = List.copyOf(
                Objects.requireNonNull(currentYearRecords, "currentYearRecords must not be null")
        );

        EvaluationState state = simulateHistory(previousMonthSpentAmount, normalizedCurrentMonthRecords, normalizedCurrentYearRecords);
        List<AppliedBenefit> appliedBenefits = evaluateTransaction(
                amount,
                merchantName,
                merchantCategory,
                paymentTags == null ? Set.of() : paymentTags,
                previousMonthSpentAmount,
                state
        );
        if (appliedBenefits.isEmpty()) {
            return Optional.empty();
        }

        long appliedAmount = appliedBenefits.stream()
                .map(AppliedBenefit::benefitAmount)
                .mapToLong(Money::amount)
                .sum();
        long rawAmount = appliedBenefits.stream()
                .map(AppliedBenefit::rawBenefitAmount)
                .mapToLong(Money::amount)
                .sum();
        return Optional.of(new BenefitQuote(
                appliedBenefits,
                Money.won(appliedAmount),
                Money.won(rawAmount)
        ));
    }

    private EvaluationState simulateHistory(
            Money previousMonthSpentAmount,
            List<SpendingRecord> currentMonthRecords,
            List<SpendingRecord> currentYearRecords
    ) {
        EvaluationState state = new EvaluationState();
        Set<java.util.UUID> currentMonthRecordIds = currentMonthRecords.stream()
                .map(SpendingRecord::id)
                .collect(java.util.stream.Collectors.toUnmodifiableSet());
        List<SpendingRecord> orderedCurrentYearRecords = currentYearRecords.stream()
                .sorted(Comparator.comparing(SpendingRecord::spentOn).thenComparing(SpendingRecord::id))
                .toList();

        for (SpendingRecord currentYearRecord : orderedCurrentYearRecords) {
            List<AppliedBenefit> appliedBenefits = evaluateTransaction(
                    currentYearRecord.amount(),
                    currentYearRecord.merchantName(),
                    currentYearRecord.merchantCategory(),
                    currentYearRecord.paymentTags(),
                    previousMonthSpentAmount,
                    state
            );
            state.consume(appliedBenefits, currentMonthRecordIds.contains(currentYearRecord.id()));
        }

        return state;
    }

    private List<AppliedBenefit> evaluateTransaction(
            Money amount,
            String merchantName,
            String merchantCategory,
            Set<String> paymentTags,
            Money previousMonthSpentAmount,
            EvaluationState state
    ) {
        List<RuleCandidate> selectedByExclusiveGroup = selectBestRulesByExclusiveGroup(
                amount,
                merchantName,
                merchantCategory,
                paymentTags,
                previousMonthSpentAmount,
                state
        );
        if (selectedByExclusiveGroup.isEmpty()) {
            return List.of();
        }

        List<AppliedBenefit> appliedBenefits = new ArrayList<>();
        Map<String, List<RuleCandidate>> sharedGroups = new HashMap<>();

        for (RuleCandidate selectedRule : selectedByExclusiveGroup) {
            Money ruleSpecificBenefit = applyRuleSpecificLimits(selectedRule, previousMonthSpentAmount, state);
            if (ruleSpecificBenefit.amount() == 0) {
                continue;
            }

            if (selectedRule.rule().sharedLimitGroupId() == null) {
                appliedBenefits.add(new AppliedBenefit(selectedRule.rule(), ruleSpecificBenefit, selectedRule.rawBenefitAmount()));
                continue;
            }

            sharedGroups.computeIfAbsent(selectedRule.rule().sharedLimitGroupId(), ignored -> new ArrayList<>())
                    .add(selectedRule.withRuleSpecificBenefit(ruleSpecificBenefit));
        }

        for (List<RuleCandidate> sharedGroupCandidates : sharedGroups.values()) {
            appliedBenefits.addAll(applySharedGroupLimits(sharedGroupCandidates, previousMonthSpentAmount, state));
        }

        return appliedBenefits;
    }

    private List<RuleCandidate> selectBestRulesByExclusiveGroup(
            Money amount,
            String merchantName,
            String merchantCategory,
            Set<String> paymentTags,
            Money previousMonthSpentAmount,
            EvaluationState state
    ) {
        Map<String, RuleCandidate> bestByExclusiveGroup = new HashMap<>();
        for (BenefitRule benefitRule : benefitRules) {
            if (!benefitRule.isEligible(
                    amount,
                    merchantName,
                    merchantCategory,
                    paymentTags,
                    previousMonthSpentAmount,
                    state.monthlyRuleCount(benefitRule),
                    state.yearlyRuleCount(benefitRule)
            )) {
                continue;
            }

            Money rawBenefitAmount = benefitRule.estimateRawBenefit(amount);
            if (rawBenefitAmount.amount() == 0) {
                continue;
            }

            RuleCandidate candidate = new RuleCandidate(benefitRule, rawBenefitAmount, Money.ZERO);
            bestByExclusiveGroup.merge(
                    benefitRule.exclusiveGroupId(),
                    candidate,
                    (left, right) -> left.rawBenefitAmount().amount() >= right.rawBenefitAmount().amount() ? left : right
            );
        }
        return List.copyOf(bestByExclusiveGroup.values());
    }

    private Money applyRuleSpecificLimits(RuleCandidate candidate, Money previousMonthSpentAmount, EvaluationState state) {
        BenefitRule benefitRule = candidate.rule();
        long appliedBenefit = candidate.rawBenefitAmount().amount();

        Money monthlyCap = benefitRule.monthlyCapFor(previousMonthSpentAmount);
        if (monthlyCap.amount() > 0) {
            long remainingMonthlyCap = Math.max(0L, monthlyCap.amount() - state.monthlyRuleBenefitUsed(benefitRule).amount());
            appliedBenefit = Math.min(appliedBenefit, remainingMonthlyCap);
        }

        if (benefitRule.yearlyBenefitCap().amount() > 0) {
            long remainingYearlyCap = Math.max(0L, benefitRule.yearlyBenefitCap().amount() - state.yearlyRuleBenefitUsed(benefitRule).amount());
            appliedBenefit = Math.min(appliedBenefit, remainingYearlyCap);
        }

        return Money.won(Math.max(0L, appliedBenefit));
    }

    private List<AppliedBenefit> applySharedGroupLimits(
            List<RuleCandidate> sharedGroupCandidates,
            Money previousMonthSpentAmount,
            EvaluationState state
    ) {
        if (sharedGroupCandidates.isEmpty()) {
            return List.of();
        }

        BenefitRule groupRule = sharedGroupCandidates.get(0).rule();
        long remainingSharedAmount = Long.MAX_VALUE;
        Money sharedMonthlyCap = groupRule.sharedMonthlyCapFor(previousMonthSpentAmount);
        if (sharedMonthlyCap.amount() > 0) {
            remainingSharedAmount = Math.min(
                    remainingSharedAmount,
                    Math.max(0L, sharedMonthlyCap.amount() - state.monthlySharedBenefitUsed(groupRule.sharedLimitGroupId()).amount())
            );
        }
        if (groupRule.sharedYearlyBenefitCap().amount() > 0) {
            remainingSharedAmount = Math.min(
                    remainingSharedAmount,
                    Math.max(0L, groupRule.sharedYearlyBenefitCap().amount() - state.yearlySharedBenefitUsed(groupRule.sharedLimitGroupId()).amount())
            );
        }

        List<RuleCandidate> orderedCandidates = sharedGroupCandidates.stream()
                .sorted(Comparator.comparingLong((RuleCandidate candidate) -> candidate.ruleSpecificBenefit().amount()).reversed())
                .toList();
        List<AppliedBenefit> appliedBenefits = new ArrayList<>();

        for (RuleCandidate orderedCandidate : orderedCandidates) {
            if (remainingSharedAmount == 0) {
                break;
            }

            long appliedAmount = orderedCandidate.ruleSpecificBenefit().amount();
            if (remainingSharedAmount != Long.MAX_VALUE) {
                appliedAmount = Math.min(appliedAmount, remainingSharedAmount);
            }
            if (appliedAmount == 0) {
                continue;
            }

            appliedBenefits.add(new AppliedBenefit(
                    orderedCandidate.rule(),
                    Money.won(appliedAmount),
                    orderedCandidate.rawBenefitAmount()
            ));
            if (remainingSharedAmount != Long.MAX_VALUE) {
                remainingSharedAmount -= appliedAmount;
            }
        }

        return appliedBenefits;
    }

    private List<PerformanceTier> sortedTiers() {
        return tiers.stream()
                .sorted(Comparator.comparingLong(tier -> tier.targetAmount().amount()))
                .toList();
    }

    private record RuleCandidate(BenefitRule rule, Money rawBenefitAmount, Money ruleSpecificBenefit) {
        private RuleCandidate withRuleSpecificBenefit(Money ruleSpecificBenefit) {
            return new RuleCandidate(rule, rawBenefitAmount, ruleSpecificBenefit);
        }
    }

    private static final class EvaluationState {
        private final Map<String, Money> monthlyRuleBenefitUsed = new HashMap<>();
        private final Map<String, Money> yearlyRuleBenefitUsed = new HashMap<>();
        private final Map<String, Integer> monthlyRuleCountUsed = new HashMap<>();
        private final Map<String, Integer> yearlyRuleCountUsed = new HashMap<>();
        private final Map<String, Money> monthlySharedBenefitUsed = new HashMap<>();
        private final Map<String, Money> yearlySharedBenefitUsed = new HashMap<>();

        private int monthlyRuleCount(BenefitRule benefitRule) {
            return monthlyRuleCountUsed.getOrDefault(benefitRule.ruleId(), 0);
        }

        private int yearlyRuleCount(BenefitRule benefitRule) {
            return yearlyRuleCountUsed.getOrDefault(benefitRule.ruleId(), 0);
        }

        private Money monthlyRuleBenefitUsed(BenefitRule benefitRule) {
            return monthlyRuleBenefitUsed.getOrDefault(benefitRule.ruleId(), Money.ZERO);
        }

        private Money yearlyRuleBenefitUsed(BenefitRule benefitRule) {
            return yearlyRuleBenefitUsed.getOrDefault(benefitRule.ruleId(), Money.ZERO);
        }

        private Money monthlySharedBenefitUsed(String sharedLimitGroupId) {
            return monthlySharedBenefitUsed.getOrDefault(sharedLimitGroupId, Money.ZERO);
        }

        private Money yearlySharedBenefitUsed(String sharedLimitGroupId) {
            return yearlySharedBenefitUsed.getOrDefault(sharedLimitGroupId, Money.ZERO);
        }

        private void consume(List<AppliedBenefit> appliedBenefits, boolean isCurrentMonthRecord) {
            Set<String> monthlyCountedRules = new HashSet<>();
            Set<String> yearlyCountedRules = new HashSet<>();

            for (AppliedBenefit appliedBenefit : appliedBenefits) {
                BenefitRule benefitRule = appliedBenefit.benefitRule();
                yearlyRuleBenefitUsed.merge(benefitRule.ruleId(), appliedBenefit.benefitAmount(), Money::add);
                yearlyCountedRules.add(benefitRule.ruleId());

                if (benefitRule.sharedLimitGroupId() != null) {
                    yearlySharedBenefitUsed.merge(benefitRule.sharedLimitGroupId(), appliedBenefit.benefitAmount(), Money::add);
                }

                if (isCurrentMonthRecord) {
                    monthlyRuleBenefitUsed.merge(benefitRule.ruleId(), appliedBenefit.benefitAmount(), Money::add);
                    monthlyCountedRules.add(benefitRule.ruleId());
                    if (benefitRule.sharedLimitGroupId() != null) {
                        monthlySharedBenefitUsed.merge(benefitRule.sharedLimitGroupId(), appliedBenefit.benefitAmount(), Money::add);
                    }
                }
            }

            for (String countedRule : yearlyCountedRules) {
                yearlyRuleCountUsed.merge(countedRule, 1, Integer::sum);
            }
            if (isCurrentMonthRecord) {
                for (String countedRule : monthlyCountedRules) {
                    monthlyRuleCountUsed.merge(countedRule, 1, Integer::sum);
                }
            }
        }
    }
}

File: card-core/src/main/kotlin/com/jean202/cardmizer/core/domain/CardPerformancePolicy.kt
package com.jean202.cardmizer.core.domain

import com.jean202.cardmizer.common.Money

data class CardPerformancePolicy(
    val cardId: CardId,
    val tiers: List<PerformanceTier>,
    val benefitRules: List<BenefitRule> = emptyList(),
) {
    init {
        require(tiers.isNotEmpty()) { "At least one performance tier is required" }
    }

    fun nextTier(spentAmount: Money): PerformanceTier? =
        sortedTiers().firstOrNull { it.targetAmount > spentAmount }

    fun highestTier(): PerformanceTier = sortedTiers().last()

    fun isFullyAchieved(spentAmount: Money): Boolean = nextTier(spentAmount) == null

    fun estimateBenefit(
        amount: Money,
        merchantName: String,
        merchantCategory: String,
        paymentTags: Set<String>,
        previousMonthSpentAmount: Money,
        currentMonthRecords: List<SpendingRecord>,
        currentYearRecords: List<SpendingRecord>,
    ): BenefitQuote? {
        val state = simulateHistory(previousMonthSpentAmount, currentMonthRecords, currentYearRecords)
        val appliedBenefits = evaluateTransaction(
            amount, merchantName, merchantCategory,
            paymentTags, previousMonthSpentAmount, state,
        )
        if (appliedBenefits.isEmpty()) return null

        val appliedAmount = appliedBenefits.sumOf { it.benefitAmount.amount }
        val rawAmount = appliedBenefits.sumOf { it.rawBenefitAmount.amount }
        return BenefitQuote(appliedBenefits, Money.won(appliedAmount), Money.won(rawAmount))
    }

    private fun simulateHistory(
        previousMonthSpentAmount: Money,
        currentMonthRecords: List<SpendingRecord>,
        currentYearRecords: List<SpendingRecord>,
    ): EvaluationState {
        val state = EvaluationState()
        val currentMonthIds = currentMonthRecords.map { it.id }.toSet()
        val ordered = currentYearRecords.sortedWith(compareBy({ it.spentOn }, { it.id }))

        for (record in ordered) {
            val applied = evaluateTransaction(
                record.amount, record.merchantName, record.merchantCategory,
                record.paymentTags, previousMonthSpentAmount, state,
            )
            state.consume(applied, record.id in currentMonthIds)
        }
        return state
    }

    private fun evaluateTransaction(
        amount: Money,
        merchantName: String,
        merchantCategory: String,
        paymentTags: Set<String>,
        previousMonthSpentAmount: Money,
        state: EvaluationState,
    ): List<AppliedBenefit> {
        val selectedByGroup = selectBestByExclusiveGroup(
            amount, merchantName, merchantCategory, paymentTags, previousMonthSpentAmount, state,
        )
        if (selectedByGroup.isEmpty()) return emptyList()

        val appliedBenefits = mutableListOf<AppliedBenefit>()
        val sharedGroups = mutableMapOf<String, MutableList<RuleCandidate>>()

        for (candidate in selectedByGroup) {
            val ruleSpecific = applyRuleSpecificLimits(candidate, previousMonthSpentAmount, state)
            if (ruleSpecific.amount == 0L) continue

            val groupId = candidate.rule.sharedLimitGroupId
            if (groupId == null) {
                appliedBenefits += AppliedBenefit(candidate.rule, ruleSpecific, candidate.rawBenefitAmount)
            } else {
                sharedGroups.getOrPut(groupId) { mutableListOf() } += candidate.withRuleSpecificBenefit(ruleSpecific)
            }
        }

        for (groupCandidates in sharedGroups.values) {
            appliedBenefits += applySharedGroupLimits(groupCandidates, previousMonthSpentAmount, state)
        }
        return appliedBenefits
    }

    private fun selectBestByExclusiveGroup(
        amount: Money,
        merchantName: String,
        merchantCategory: String,
        paymentTags: Set<String>,
        previousMonthSpentAmount: Money,
        state: EvaluationState,
    ): List<RuleCandidate> {
        val bestByGroup = mutableMapOf<String, RuleCandidate>()
        for (rule in benefitRules) {
            if (!rule.isEligible(
                    amount, merchantName, merchantCategory, paymentTags,
                    previousMonthSpentAmount, state.monthlyRuleCount(rule), state.yearlyRuleCount(rule),
                )
            ) continue

            val raw = rule.estimateRawBenefit(amount)
            if (raw.amount == 0L) continue

            val candidate = RuleCandidate(rule, raw, Money.ZERO)
            bestByGroup.merge(rule.exclusiveGroupId, candidate) { left, right ->
                if (left.rawBenefitAmount.amount >= right.rawBenefitAmount.amount) left else right
            }
        }
        return bestByGroup.values.toList()
    }

    private fun applyRuleSpecificLimits(
        candidate: RuleCandidate,
        previousMonthSpentAmount: Money,
        state: EvaluationState,
    ): Money {
        var applied = candidate.rawBenefitAmount.amount
        val monthlyCap = candidate.rule.monthlyCapFor(previousMonthSpentAmount)
        if (monthlyCap.amount > 0) {
            val remaining = maxOf(0L, monthlyCap.amount - state.monthlyRuleBenefitUsed(candidate.rule).amount)
            applied = minOf(applied, remaining)
        }
        if (candidate.rule.yearlyBenefitCap.amount > 0) {
            val remaining = maxOf(0L, candidate.rule.yearlyBenefitCap.amount - state.yearlyRuleBenefitUsed(candidate.rule).amount)
            applied = minOf(applied, remaining)
        }
        return Money.won(maxOf(0L, applied))
    }

    private fun applySharedGroupLimits(
        candidates: List<RuleCandidate>,
        previousMonthSpentAmount: Money,
        state: EvaluationState,
    ): List<AppliedBenefit> {
        if (candidates.isEmpty()) return emptyList()
        val groupRule = candidates.first().rule
        var remaining = Long.MAX_VALUE

        val sharedMonthlyCap = groupRule.sharedMonthlyCapFor(previousMonthSpentAmount)
        if (sharedMonthlyCap.amount > 0) {
            remaining = minOf(remaining, maxOf(0L, sharedMonthlyCap.amount - state.monthlySharedBenefitUsed(groupRule.sharedLimitGroupId!!).amount))
        }
        if (groupRule.sharedYearlyBenefitCap.amount > 0) {
            remaining = minOf(remaining, maxOf(0L, groupRule.sharedYearlyBenefitCap.amount - state.yearlySharedBenefitUsed(groupRule.sharedLimitGroupId!!).amount))
        }

        val applied = mutableListOf<AppliedBenefit>()
        for (candidate in candidates.sortedByDescending { it.ruleSpecificBenefit.amount }) {
            if (remaining == 0L) break
            var amount = candidate.ruleSpecificBenefit.amount
            if (remaining != Long.MAX_VALUE) amount = minOf(amount, remaining)
            if (amount == 0L) continue
            applied += AppliedBenefit(candidate.rule, Money.won(amount), candidate.rawBenefitAmount)
            if (remaining != Long.MAX_VALUE) remaining -= amount
        }
        return applied
    }

    private fun sortedTiers(): List<PerformanceTier> =
        tiers.sortedBy { it.targetAmount.amount }

    private data class RuleCandidate(
        val rule: BenefitRule,
        val rawBenefitAmount: Money,
        val ruleSpecificBenefit: Money,
    ) {
        fun withRuleSpecificBenefit(benefit: Money) = copy(ruleSpecificBenefit = benefit)
    }

    private class EvaluationState {
        private val monthlyRuleBenefit = mutableMapOf<String, Money>()
        private val yearlyRuleBenefit = mutableMapOf<String, Money>()
        private val monthlyRuleCount = mutableMapOf<String, Int>()
        private val yearlyRuleCount = mutableMapOf<String, Int>()
        private val monthlySharedBenefit = mutableMapOf<String, Money>()
        private val yearlySharedBenefit = mutableMapOf<String, Money>()

        fun monthlyRuleCount(rule: BenefitRule) = monthlyRuleCount.getOrDefault(rule.ruleId, 0)
        fun yearlyRuleCount(rule: BenefitRule) = yearlyRuleCount.getOrDefault(rule.ruleId, 0)
        fun monthlyRuleBenefitUsed(rule: BenefitRule) = monthlyRuleBenefit.getOrDefault(rule.ruleId, Money.ZERO)
        fun yearlyRuleBenefitUsed(rule: BenefitRule) = yearlyRuleBenefit.getOrDefault(rule.ruleId, Money.ZERO)
        fun monthlySharedBenefitUsed(groupId: String) = monthlySharedBenefit.getOrDefault(groupId, Money.ZERO)
        fun yearlySharedBenefitUsed(groupId: String) = yearlySharedBenefit.getOrDefault(groupId, Money.ZERO)

        fun consume(appliedBenefits: List<AppliedBenefit>, isCurrentMonth: Boolean) {
            val yearlyCountedRules = mutableSetOf<String>()
            val monthlyCountedRules = mutableSetOf<String>()

            for (benefit in appliedBenefits) {
                val ruleId = benefit.benefitRule.ruleId
                yearlyRuleBenefit.merge(ruleId, benefit.benefitAmount, Money::add)
                yearlyCountedRules += ruleId

                benefit.benefitRule.sharedLimitGroupId?.let { groupId ->
                    yearlySharedBenefit.merge(groupId, benefit.benefitAmount, Money::add)
                }

                if (isCurrentMonth) {
                    monthlyRuleBenefit.merge(ruleId, benefit.benefitAmount, Money::add)
                    monthlyCountedRules += ruleId
                    benefit.benefitRule.sharedLimitGroupId?.let { groupId ->
                        monthlySharedBenefit.merge(groupId, benefit.benefitAmount, Money::add)
                    }
                }
            }
            yearlyCountedRules.forEach { yearlyRuleCount.merge(it, 1, Int::plus) }
            if (isCurrentMonth) monthlyCountedRules.forEach { monthlyRuleCount.merge(it, 1, Int::plus) }
        }
    }
}
